### DISCUSSION TITLE
Issue to bring records with their relations

### DISCUSSION THREAD

**Discussion by Graffalo92** [pocketbase/pocketbase#3193] [Q&A] [answered]
const data = await pb.collection("destinations").getFullList({
      expand: "metadata_id",
      sort: "+created,route_name",
      filter: "is_active = true",
    });
    
    I get the collection with their relationships. The thing is that return response of relation is like this 
    expand: { metadata_id: [Record] },
    
    as a result it seems that i cant access. Am i missing something ?

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

To access the expanded relation, you can do something like: `data[0].expand.metadata_id[0].id`.
The expanded value could be either an array or plain record depending on the relation type - multiple or single.

If that's not what you are looking for, please elaborate.

**Comment by Graffalo92** [comment]
Thank you for answering so fast.  I am trying to console.log the response coming from data. Destinations has single relarion with metadata_destination collection under destinations.metadata_id. The response i get for expand is this expand: { metadata_id: [Record] }.
As a result i cant acces it.

**Reply by ganigeorgiev** [reply]
Sorry, I'm still failing to understand what do you mean.

Please elaborate what do you mean by _"As a result i cant acces it"_.

**Comment by Graffalo92** [comment]
I literally get as a response 
Record {
    collectionId: 'n6ijgf9xbnic7vu',
    collectionName: 'destinations',
    created: '2023-08-04 08:37:56.927Z',
    destination_name: { en: 'Athens', es: 'Atenas', pt: 'Atenas' },
    expand: { metadata_id: [Record] },
    id: 'c3x9ungwz4fonvf',
    is_active: true,
    metadata_id: '5wq0od50sna08zv',
rest...

**Reply by ganigeorgiev** [reply]
What is wrong with the above response?

What is the expected output?

**Reply by Graffalo92** [reply]
Shouldn't i get the record expanded in order to access the data of the relation ?

**Reply by ganigeorgiev** [reply]
I'm confused. Based on the above output it is there - under the `expand.metadata_id` key. If you expect the expanded record to be under `metadata_id` then this is not how it works as we don't modify the actual record field values.

**Reply by Graffalo92** [reply]
Yeah i see that in my console.log but i dont get any data. I just get [Record]. If i console.log data.expand.metadata_id** i get nothing "Undefined".  Thats what i am trying to undestand. If  the way i am trying to do this dont actually bring the data of the relational collection.

**Reply by ganigeorgiev** [reply]
Without a more complete code sample I'm not sure how to help. If the record is there when using `console.log` then you should be able to access it and it definitely shouldn't be `undefined`.

**Reply by ganigeorgiev** [reply]
To rule out some serialization issue (in case you use a SSR framework) you can try to update your JS SDK dependency to `0.17.0-rc1` in your package.json and try again since in the rc version we no longer wrap the response in custom classes and return POJOs (see https://github.com/pocketbase/pocketbase/discussions/3134).

**Reply by Graffalo92** [reply]
I already did that. In the case with version 0.17 i get expand: { metadata_id: [Object] }, instead of just [Record]. yes i am using next js 13.  For single record fetch i can normally see the data of the relation.

**Reply by ganigeorgiev** [reply]
I really don't know how to help you.

Again, if you have access to the record when calling `console.log` then you should be able to access the object fields just fine.

Double check that you are not hitting the default Next.js cache - see https://github.com/pocketbase/pocketbase/discussions/2695.

To rule out PocketBase and JS SDK issue, you can try to create a static index.html file inside `pb_public/index.html`:

```html
<!DOCTYPE html />
<html>
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <title>Test</title>
    </head>
    <body>
        <script type="module">
            import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@develop/dist/pocketbase.es.mjs";

            const pb = new PocketBase("http://127.0.0.1:8090");

            const data = await pb.collection("destinations").getFullList({
              expand: "metadata_id",
              sort: "+created,route_name",
              filter: "is_active = true",
            });

            console.log(data)
        </script>
    </body>
</html>
```
Open `http://127.0.0.1:8090` and check the console.log output. If you see the expand data then it is most likely a Next.js cache/configuration issue.

**Reply by Graffalo92** [reply]
Something is wrong with next js then. even though i use export const fetchCache = "default-no-store"; i literally still get [Record] or [Object] (with 0.17) and not the actuall data of the relationship record. In the code you provided it works fine. 

     Data i expect
            "expand": {
                "metadata_id": {
                    "collectionId": "x9sh8wrkx5p022y",
                    "collectionName": "metadata_destination",
                    "created": "2023-08-25 14:48:53.831Z",
                    "id": "77iaa486je6kv4c",
                    "link": {
                        "en": "Athens-City-An-Ageless-Odyssey-of-Activities ",
                        "es": "Atenas-Ciudad-Una-Odisea-Intemporal-de-Actividades",
                        "pt": "Atenas-Cidade-Uma-Odisseia-Eterna-de-Atividades"
                    },
                    "updated": "2023-08-25 14:48:53.831Z"
                }
            },
            
        Data i get with next js 13 
        expand: { metadata_id: [Object] },

**Reply by ganigeorgiev** [reply]
> i literally still get [Record] or [Object] (with 0.17)

Without a code sample how you actually access the data, I'm out of ideas.

The above seems to be the string representation of the data, **not the actual object value**. Try to add `console.log(typeof record.expand.metadata_id)` (_replace `record` with one of the `getFullList` entries_) and see what is the output.

In any case, as tested per https://github.com/pocketbase/pocketbase/discussions/3193#discussioncomment-6823956, I don't think this is a PocketBase or JS SDK issue. If you think otherwise, please create a **minimal** reproducible repo and I'll try to test it.

**Comment by Graffalo92** [comment]
console.log(typeof record.expand.metadata_id)
TypeError: Cannot read properties of undefined (reading 'metadata_id')

Here is a mini repo. of my page.js. Thats my code to the actual source. 
https://github.com/Graffalo92/testpocket

With fresh pocketbase install and next js 13 app with 2 collections and 1 relationship this is the response i get.

        [
          Record {
            collectionId: 't9jmmiw0oq1f1ye',
            collectionName: 'destinations',
            created: '2023-08-25 15:28:02.968Z',
            expand: { metadata_id: [Record] },
            field: 'test',
            id: '4vkd2teislbe1gc',
            metadata_id: 'qna0xqz891zwcai',
            updated: '2023-08-25 15:28:40.515Z'
          }
        ]
        
 If i run the same from the code you provided
 
     [
        {
            "collectionId": "t9jmmiw0oq1f1ye",
            "collectionName": "destinations",
            "created": "2023-08-25 15:28:02.968Z",
            "expand": {
                "metadata_id": {
                    "collectionId": "6ojlrq5bp8l9k5i",
                    "collectionName": "metadata",
                    "created": "2023-08-25 15:28:21.155Z",
                    "id": "qna0xqz891zwcai",
                    "text": "testre",
                    "updated": "2023-08-25 15:28:21.155Z"
                }
            },
            "field": "test",
            "id": "4vkd2teislbe1gc",
            "metadata_id": "qna0xqz891zwcai",
            "updated": "2023-08-25 15:28:40.515Z"
        }
    ]

**Reply by ganigeorgiev** [reply]
In https://github.com/Graffalo92/testpocket/blob/master/app/page.js#L6 you are not returning anything from the function and it resolves to empty `Promise`, aka. add `return data` after the `console.log(data)` call.

**Reply by Graffalo92** [reply]
Even if i return data from getData() function. And console.log in page Home() function is still the same result.

**Reply by ganigeorgiev** [reply, marked as answer]
I really don't know how to explain it better.

The `console.log` is just a string representation of the object and it could differ depending where the output will be - in the browser console, stdout from node, etc.

If you really want to compare the results you can serialize the object with `JSON.stringify(data)` and verify that it is there, but I'm failing to understand what is actually the issue as the above repo doesn't show how you actually access the data.

I'm sorry but I don't think I can help you further as it seems that we are going in circles.
